judger/src/tester/spj.rs

```diff
@@ -0,0 +1,383 @@
+    ///
+    /// This function is in fact **not** async, but making it async forces it to
+    /// be run on some async environment, so [`tokio::runtime::Handle::current`].
+    /// would run without problem.
+    pub async fn spawn_futures(&self) -> JoinHandle<()> {
+        self.rt.spawn_executor(TokioSpawner(Handle::current()))
+    }
+
+    /// Callback for initializing special judge
+    pub async fn spj_global_init(&self, config: &JudgerPublicConfig) -> anyhow::Result<()> {
+        run_promise_like!(self.ctx, SPJ_INIT_FN, (config,), |x| x).map_err(|e| e.into())
+    }
+
+    /// Callback for mapping exec
+    pub async fn spj_map_exec(&self, config: &[RawStep]) -> anyhow::Result<Vec<RawStep>> {
+        run_promise_like!(self.ctx, SPJ_TRANSFORM_FN, (config,), |x| x).map_err(|e| e.into())
+    }
+
+    /// Callback for case init
+    pub async fn spj_case_init(
+        &self,
+        case: &TestCase,
+        mappings: &HashMap<String, String>,
+    ) -> anyhow::Result<()> {
+        run_promise_like!(self.ctx, SPJ_CASE_INIT_FN, (case, mappings), |x| x).map_err(|e| e.into())
+    }
+
+    /// Callback for case judging
+    pub async fn spj_case_judge(&self, results: &[ProcessInfo]) -> anyhow::Result<SpjResult> {
+        run_promise_like!(self.ctx, SPJ_CASE_FN, (results,), |x| x).map_err(|e| e.into())
+    }
+
+    fn detect_features(&self) -> SpjFeatures {
+        self.ctx.with(|ctx| {
+            let globals = ctx.globals();
+            let global_init = globals
+                .get::<_, rquickjs::Value>(SPJ_INIT_FN)
+                .map_or(false, |v| v.is_function());
+            let transform_exec = globals
+                .get::<_, rquickjs::Value>(SPJ_TRANSFORM_FN)
+                .map_or(false, |v| v.is_function());
+            let case_init = globals
+                .get::<_, rquickjs::Value>(SPJ_CASE_INIT_FN)
+                .map_or(false, |v| v.is_function());
+            let case = globals
+                .get::<_, rquickjs::Value>(SPJ_CASE_FN)
+                .map_or(false, |v| v.is_function());
+            SpjFeatures {
+                global_init,
+                transform_exec,
+                case_init,
+                case,
+            }
+        })
+    }
+}
+
+/// Make a special judger using the given script path.
+pub async fn make_spj(script_path: &Path) -> anyhow::Result<SpjEnvironment> {
+    let mut spj = SpjEnvironment::new().context("when creating spj environment")?;
+    let script = tokio::fs::read(script_path)
+        .await
+        .context("when reading spj input file")?;
+    let script = String::from_utf8_lossy(&script);
+    spj.load_script(&script)
+        .context("when loading spj script")?;
+    Ok(spj)
+}
+
+#[derive(Debug, FromJs)]
+#[quickjs(rename_all = "camelCase")]
+pub struct SpjResult {
+    pub accepted: bool,
+    pub score: Option<f64>,
+    pub reason: Option<String>,
+}
+
+/// Represents enabled features of a special judge instance.
```

Comment:
Would you mind specifying these features?

Reply:
Done.